Add Combined Energy integration

timsavage opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
New integration of [Combined Energy](http://combined.energy/) API for home energy management system.

Provides energy/power/temperature sensors for supported devices. The integration domain is `combined_energy`, I briefly named it `combined_energy_api` however, that didn't seem consistent with other integration domain names.

Client library release: https://github.com/timsavage/combined-energy-api/tree/0.5

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/24856

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`. - All files are tested to 100% coverage

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

timsavage committed:
Define constants for component

timsavage committed:
Add module init

timsavage committed:
Manifest for new component

timsavage committed:
Added Combined Energy brand

timsavage committed:
Updated config flow to handle more errors and update strings

timsavage committed:
Add shim of aiter/anext for python 3.9

timsavage committed:
Setup API client

timsavage committed:
Added and setup coordinators for connection and readings

timsavage committed:
Store installation for future use determining devices and entities

timsavage committed:
Reduce duplication in error handling

timsavage committed:
Simplify data service and just return devices with ID's

timsavage committed:
Bump version of api client to 0.4

timsavage committed:
Don't much the user defined name

timsavage committed:
Added more sensor types

timsavage committed:
Specify initial_delta to reduce the size of the first read

timsavage committed:
Add factories to generate all sensors and devices

timsavage committed:
Include strict typing

timsavage committed:
Renamed back to combined_energy

timsavage committed:
Renamed back to combined_energy

timsavage committed:
Defined sensors

timsavage committed:
Split sensors into different types

timsavage committed:
Add a 15 minute delayed log session refresh call

timsavage committed:
Renamed to combined_energy

timsavage committed:
Re-organise power/energy sensors, disable more by default.

timsavage committed:
Re-organise power/energy sensors, disable more by default.

timsavage committed:
Add listener to ensure log session is updated

timsavage committed:
Remove un-needed logger

timsavage committed:
Use extra_state_attributes to return the last_updated value in place of last_reset reducing log noise

timsavage committed:
Correct units of power factor and voltage measurements

timsavage committed:
Define last_reset using a property

timsavage committed:
Add handling for config failures, add test cases

timsavage committed:
Add tests for coordinator and remove path that can't actually occur

timsavage committed:
Removed unused method

timsavage committed:
Add tests for Connected Sensor

timsavage committed:
Initial test of config flow

timsavage committed:
Full test coverage for config flow

timsavage committed:
Some tweaks to design of sensors to make easier to test, tests for sensors

timsavage committed:
Update duplicate check to use _abort_if_unique_id_configured

timsavage committed:
Bump sensor coverage to 100%

timsavage committed:
Tidy up and re-use existing variables extracted from user_input dict

timsavage committed:
Update manifest.json with quality info

matthewsachse commented:
I've had  this component running in my HA instance with good success. Entities all updating from the Combined Energy portal as well as feeding the HA energy dashboard.

timsavage commented:
I have now had this operating on my own instance (on a Yellow) without error over the past three months.

Lash-L commented on homeassistant/components/combined_energy/config_flow.py:
```diff
@@ -0,0 +1,107 @@
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        self._errors = {}
```
Why are you creating a private member variable here?

timsavage replied:
This is a variable being cleared, not defined. It's defined in the `__init__`

Lash-L replied:
For some reason i thought it was self.errors in the init.

That being said, I believe one of the linters still views this as defining.

I think you should do self._errors.clear()

timsavage replied:
Linter had no issue with this. I prefer to use a new dict when clearing.

Lash-L replied:
Yep you're right - not sure what the error was i had gotten in the past doing the same thing. But this is all good

Lash-L commented on homeassistant/components/combined_energy/config_flow.py:
```diff
@@ -0,0 +1,107 @@
+            password = user_input[CONF_PASSWORD]
+            installation_id = user_input[CONF_INSTALLATION_ID]
+
+            await self.async_set_unique_id(f"{DOMAIN}-{installation_id}")
```
I believe it can just be installation id - depending on what installation id is https://developers.home-assistant.io/docs/config_entries_config_flow_handler#unique-id-requirements

timsavage replied:
The installation ID is generated by the combined energy API. Safer to prefix it with a known value to ensure it's uniqueness.

Lash-L replied:
My understanding of unique ids is that they are already domain specific. so the DOMAIN- tag is unneeded.

Lash-L commented on homeassistant/components/combined_energy/const.py:
```diff
@@ -0,0 +1,337 @@
+            state_class=SensorStateClass.MEASUREMENT,
+            native_unit_of_measurement="%",
+            device_class=SensorDeviceClass.POWER_FACTOR,
+            entity_registry_enabled_default=False,
+        ),
+        SensorEntityDescription(
+            key="power_factor_c",
+            name="Power Factor C",
+            state_class=SensorStateClass.MEASUREMENT,
+            native_unit_of_measurement="%",
+            device_class=SensorDeviceClass.POWER_FACTOR,
+            entity_registry_enabled_default=False,
+        ),
+        SensorEntityDescription(
+            key="voltage_a",
+            name="Voltage A",
+            state_class=SensorStateClass.MEASUREMENT,
+            native_unit_of_measurement="V",
+            device_class=SensorDeviceClass.VOLTAGE,
+        ),
+        SensorEntityDescription(
+            key="voltage_b",
+            name="Voltage B",
+            state_class=SensorStateClass.MEASUREMENT,
+            native_unit_of_measurement="V",
+            device_class=SensorDeviceClass.VOLTAGE,
+            entity_registry_enabled_default=False,
+        ),
+        SensorEntityDescription(
+            key="voltage_c",
+            name="Voltage C",
+            state_class=SensorStateClass.MEASUREMENT,
+            native_unit_of_measurement="V",
+            device_class=SensorDeviceClass.VOLTAGE,
+            entity_registry_enabled_default=False,
+        ),
+    ],
+    "GENERIC_CONSUMER": SENSOR_DESCRIPTIONS_GENERIC_CONSUMER,
+    "ENERGY_BALANCE": SENSOR_DESCRIPTIONS_GENERIC_CONSUMER,
+}
```
I'm not 100% but I don't think descriptions should go in const, I think they should stay in sensor.py

Lash-L replied:
Looking at the codebase again, it does happen a few times, it's just not super common, so this is probably okay

timsavage replied:
This pattern was taken from the Solar Edge API. 
[Supported overview sensors](https://github.com/home-assistant/core/blob/d106cb48d2cebfc41883e9faf51b2769b8b5dffb/homeassistant/components/solaredge/const.py#L29)

Lash-L commented on homeassistant/components/combined_energy/coordinator.py:
```diff
@@ -0,0 +1,136 @@
+from datetime import timedelta
+from typing import Generic, TypeVar
+
+from combined_energy import CombinedEnergy
+from combined_energy.exceptions import CombinedEnergyAuthError, CombinedEnergyError
+from combined_energy.helpers import ReadingsIterator
+from combined_energy.models import ConnectionStatus, DeviceReadings
```
make these local imports
from .
from .exceptions

timsavage replied:
`combined_energy` is not a local package. It is the external async client library for combined energy.
https://github.com/timsavage/combined-energy-api

Lash-L replied:
Ah, my bad, I took too quick of a glance at it.

Lash-L commented on homeassistant/components/combined_energy/coordinator.py:
```diff
@@ -0,0 +1,136 @@
+_T = TypeVar("_T")
+
+
+class CombinedEnergyDataService(Generic[_T]):
```
I'm a bit confused, why are you not just using a coordinator object? Feels like you're over complicating this. I could be missing something

timsavage replied:
This is due to the underlaying API, which is actually multiple data services.

Since they all utilise the same basic pattern I provide that in a generic base class that each service inherits to provide the concrete implementation.

Lash-L commented on homeassistant/components/combined_energy/manifest.json:
```diff
@@ -0,0 +1,11 @@
+{
+  "domain": "combined_energy",
+  "name": "Combined Energy",
+  "integration_type": "hub",
+  "documentation": "https://www.home-assistant.io/integrations/combined_energy",
+  "requirements": ["combined-energy-api==0.5"],
+  "codeowners": ["@timsavage"],
+  "iot_class": "cloud_polling",
+  "quality_scale": "silver",
+  "config_flow": true
```
should be in domain, name, alphabetical order

timsavage replied:
Is there documentation specifying the order? I'm not able to find any mention of it.

Order is irrelevant in JSON for object types. I'd be more inclined to order in the same order as the documentation lists the values.

Lash-L replied:
It isn't actually written down but it should be.

https://github.com/home-assistant/core/pull/87020

Run pre-commit and it should probably notify you

timsavage replied:
Ah was added after I opened the PR.

Lash-L commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,270 @@
+    async_add_entities(entities)
+
+
+class CombinedEnergyConnectedSensor(CoordinatorEntity, BinarySensorEntity):
```
CoordinatorEntity[YourCoordinatorName] is ideal, but I'm not sure how that works with your services system you built

Lash-L replied:
Also this is a binary sensor, you can't have this in here/ your initial PR. 

PRs are limited to one platform, aka only Sensors or only Binary Sensors

timsavage replied:
This is an important piece of state information that indicates if the monitoring hardware is connected to the Combined Energy monitoring platform. I don't have a view on how to exclude this entity, would prefer to defer to a maintainer's opinion.

Lash-L replied:
Absolutely feel free to defer for their opinion, but from what I've seen it is a pretty strong rule they hold to. Typically it is also in a seperate file binary_sensor.py where you can set up the BinarySensor platform separately from the non-binary sensors.

But absolutely, feel free to defer to them

Lash-L commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,270 @@
+        self.installation = installation
+        self.readings = readings
+
+    def _generate_device_info(self, device: Device) -> DeviceInfo:
+        """Generate device info from API device response."""
+        return DeviceInfo(
+            identifiers={
+                (
+                    DOMAIN,
+                    f"install_{self.installation.installation_id}-device_{device.device_id}",
+                )
+            },
+            manufacturer=device.device_manufacturer,
+            model=device.device_model_name,
+            name=device.display_name,
+        )
```
Instead of generating device info, you should be able to just create a CombinedEnergyEntity object, do it in the init there and then make all of your other entities based off of that one

Lash-L replied:
instead of calling this code, you can create the following:

```
CombinedEnergyEntity(CoordinatorEntity):

def __init__(self,device, ...):
self._attr_device_info = DeviceInfo(
            identifiers={
                (
                    DOMAIN,
                    f"install_{self.installation.installation_id}-device_{device.device_id}",
                )
            },
            manufacturer=device.device_manufacturer,
            model=device.device_model_name,
            name=device.display_name,
        )
```

Then your CombinedEnergyReadingsSensor(CombinedEnergyEntity) and you just pass it the device as you do now and then it should be handled in the super().__init__()

timsavage replied:
The code as is creates a single `DeviceInfo` instance for each device which is then passed to each individual sensor instance.  It's a one-to-many relationship. Why create unnecessary instances of `DeviceInfo`?

Lash-L replied:
You could do it Per coordinator instead then if you'd like

timsavage replied:
That wouldn't work either. There is a single readings co-ordinator that obtains data for all the devices related to an _installation_ when polled, not a single device. On each poll of the Readings co-ordinator the integration then fans out all of the data to the various devices and sensors. The design I have used is the simplest I could provide that correctly constructs the required objects.

There is a separate LogSession Co-ordinator (Combined Energy term) that needs to be pinged periodically to keep the _log-session_, I assume this is actually a mod-bus session alive or data stops being provided until it is called again.

Much of the complexity of this integration comes from the design of the API provided by Combined Energy.

Lash-L commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,270 @@
+        return self._raw_value is not None
+
+    @abstractmethod
+    def _to_native_value(self, raw_value):
+        """Convert non-none raw value into usable sensor value."""
```
instead of doing this, you can add a mixin to your sensor_entity descriptions of a value_fn that does what ever operations you need it to. There are plenty of examples in the code

timsavage replied:
That approach doesn't work in this case. Each of the different sensor types handles the data from the API in different ways and accesses multiple member variables to do so. 
The Sensor types are selected based on the SensorDeviceClass.

Lash-L commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,270 @@
+}
+
+
+class CombinedEnergyReadingsSensorFactory:
```
I've never seen someone build a factory for this task and I'm not sure how the maintainers will feel about it.

timsavage replied:
A factory makes sense when devices need to be enumerated from the combined energy API. They are not known before this process is completed.

Lash-L commented on homeassistant/components/combined_energy/strings.json:
```diff
@@ -0,0 +1,23 @@
+    "error": {
+      "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]",
+      "invalid_auth": "[%key:common::config_flow::error::invalid_auth%]",
+      "unknown": "[%key:common::config_flow::error::unknown%]"
```
You don't have an unknown

timsavage replied:
Unknown is now used.

Lash-L commented on homeassistant/components/combined_energy/config_flow.py:
```diff
@@ -0,0 +1,107 @@
+            self._errors["base"] = "cannot_connect"
+        except CombinedEnergyError:
+            LOGGER.exception("Unexpected error verifying connection to API")
+            self._errors["base"] = "cannot_connect"
```
Is this supposed to be unknown?

timsavage replied:
Yes this should probably have a different error message to indicate to the end user.

Lash-L commented on homeassistant/components/combined_energy/translations/en.json:
Remove this file - no longer in use

timsavage replied:
It's a generated file, by hassfest

Lash-L commented on tests/components/combined_energy/test_combined_energy.py:
```diff
@@ -0,0 +1,104 @@
+    return mock_api_instance
+
+
+async def test_async_setup_entry__where_component_is_successfully_registered(
+    hass: HomeAssistant, config_entry: ConfigEntry, monkeypatch, installation
+):
```
needs return -> None

Lash-L commented on tests/components/combined_energy/test_combined_energy.py:
```diff
@@ -0,0 +1,104 @@
+        await combined_energy.async_setup_entry(hass, config_entry)
+
+
+async def test_async_setup_entry__where_connection_fails(
```
needs return -> None

Lash-L commented on tests/components/combined_energy/test_combined_energy.py:
```diff
@@ -0,0 +1,104 @@
+        installation_id=999999999,
+        session=ANY,
+    )
+    assert (
+        hass.data["combined_energy"]["beef"][combined_energy.DATA_API_CLIENT]
+        == mock_api_instance
+    )
+    assert (
+        hass.data["combined_energy"]["beef"][combined_energy.DATA_INSTALLATION]
+        is installation
+    )
```
You are not supposed to access hass.data within test
https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations

Lash-L commented on tests/components/combined_energy/test_combined_energy.py:
```diff
@@ -0,0 +1,104 @@
+
+async def test_async_unload_entry__where_component_is_successfully_unloaded(
+    hass: HomeAssistant, config_entry: ConfigEntry
+):
```
-> None

Lash-L commented on tests/components/combined_energy/test_combined_energy.py:
```diff
@@ -0,0 +1,104 @@
+    actual = await combined_energy.async_unload_entry(hass, config_entry)
+
+    assert actual is not None
+    assert hass.data["combined_energy"] == {}
```
https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations

Lash-L commented on tests/components/combined_energy/test_config_flow.py:
```diff
@@ -0,0 +1,160 @@
+from homeassistant.core import HomeAssistant
+
+
+async def test_form(hass: HomeAssistant):
```
I'm not going to continue to comment it on each one, but you need a return type on all of your functions

Lash-L reviewed (changes requested):
Here's my feedback, I'm not a maintainer, but hopefully this gets you started and gets the review process started for your integration

timsavage commented on homeassistant/components/combined_energy/config_flow.py:
```diff
@@ -0,0 +1,107 @@
+        except CombinedEnergyAuthError:
+            self._errors["base"] = "invalid_auth"
+        except CombinedEnergyPermissionError:
+            self._errors[CONF_INSTALLATION_ID] = "installation_not_accessible"
```
Yes will need to add that value.

timsavage committed:
Add return types for tests.

This doesn't really add anything to clarity of the code as all tests by definition return None!

timsavage committed:
Changed the unknown error case to use the `unknown` key.

timsavage committed:
Re-ordered by pre-commit

timsavage committed:
Update strings with missing installation_not_accessible entry

timsavage committed:
Update tests to follow Hass guidelines.

Lash-L commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,270 @@
+        return None
+
+    @property
+    def _raw_value(self):
```
Needs a return ->

Lash-L commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,270 @@
+            return device_readings.range_start
+        return None
+
+    def _to_native_value(self, raw_value):
```
All of these need a -> type and a type on raw_value

Lash-L commented on homeassistant/components/combined_energy/config_flow.py:
```diff
@@ -0,0 +1,107 @@
+
+    VERSION = 1
+
+    def __init__(self):
```
return -> None

Lash-L commented on tests/components/combined_energy/test_config_flow.py:
```diff
@@ -0,0 +1,164 @@
+    assert result["errors"] == {}
+
+    with patch(
+        "combined_energy.client.CombinedEnergy.installation", side_effect=exception
```
instead of doing a global patch, they prefer for you to patch the instance within the component. patch("homeassistant.components.

timsavage committed:
Add types to private variables (not required by linter)

timsavage committed:
Fixed bug caused by removing the DOMAIN in the installation ID

timsavage committed:
Split connected sensor into binary_sensor file and update tests

timsavage committed:
Updated calls to patch to not use global package but the local import.

timsavage commented:
> I think I've reached my limit on what I've been exposed to. Hopefully my review was helpful and gets the process started. Best of luck!

Thank you @Lash-L , there were a few things had missed along the way. Being challenged on a software design is a good thing, this was not an easy integration and this is probably the third iteration to get everything reliably working. Has been operating reliably for 5ish months now.

I'll need to go update my local instance with the changes to give them a good test. 

Cheers.

seriocomic commented:
Just an observer - looking forward to this getting approval! Nice work @timsavage

JrtPec commented on tests/components/combined_energy/test_config_flow.py:
```diff
@@ -0,0 +1,166 @@
+        )
+        await hass.async_block_till_done()
+
+    assert result2["type"] == "create_entry"
```
Checking against `FlowResultType.CREATE_ENTRY` seems better

JrtPec commented on tests/components/combined_energy/test_config_flow.py:
```diff
@@ -0,0 +1,166 @@
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    assert result["type"] == "form"
```
`assert result["type"] == FlowResultType.FORM` seems better.
(happens multiple times in this file)

JrtPec commented on tests/components/combined_energy/test_config_flow.py:
```diff
@@ -0,0 +1,166 @@
+        )
+        await hass.async_block_till_done()
+
+    assert result2["type"] == "abort"
```
Checking agains `FlowResultType.ABORT` seems better.

timsavage committed:
Change flow results to use the new FlowResultType enum

edenhaus commented on homeassistant/components/combined_energy/translations/en.json:
Please remove the translations folder. All translation strings go into the file `strings.json`

edenhaus commented on homeassistant/brands/combined_energy.json:
Please remove this file. The brand's file is only required to group two or more integrations, which support only a part of the manufacturer.

edenhaus commented on homeassistant/components/combined_energy/.strict-typing:
Please remove this file and add your integration to the [`.strict-typing`](https://developers.home-assistant.io/docs/development_typing/#including-modules-for-strict-type-checking) file in the root folder.

edenhaus reviewed (changes requested):
I have done an initial review.

Please address the requested changes and fix the CI.

timsavage committed:
Migrate to using root .strict typing file

timsavage committed:
Remove un-needed combined_energy.json in brands folder

timsavage committed:
Remove backports.

This added support for async iterators to Python 3.10. Now that hassio has standardised on Python 3.11 this is no longer required.

timsavage committed:
Remove binary_sensor

Move this sensor into a future PR.

timsavage committed:
Remove translation entry.

timsavage commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,238 @@
+"""Sensors and factory for enumerating devices from the Combined Energy API."""
```
Not sure what you mean. 

All the translatable strings are listed in `strings.json` this file does not contain any un-translateable strings...

edenhaus reviewed (changes requested):
Please make sure CI is passing

timsavage commented on homeassistant/components/combined_energy/config_flow.py:
```diff
@@ -0,0 +1,107 @@
+            password = user_input[CONF_PASSWORD]
+            installation_id = user_input[CONF_INSTALLATION_ID]
+
+            await self.async_set_unique_id(str(installation_id))
```
It is unique accross all accounts.

timsavage committed:
Update homeassistant/components/combined_energy/config_flow.py

Assume dict will always contain CONF_NAME

Co-authored-by: Robert Resch <robert@resch.dev>

timsavage committed:
Update homeassistant/components/combined_energy/config_flow.py

Don't check for None. I am assuming that HA will not pass an empty dict

Co-authored-by: Robert Resch <robert@resch.dev>

timsavage committed:
Regenerate requirements_test_all.txt

timsavage committed:
Regenerate integrations.json and mypy.ini

timsavage committed:
Removed binary sensor tests

timsavage committed:
Move sensor specific constants out of const file.

timsavage committed:
Black re-format

timsavage committed:
Remove additional logger statements.

The combined energy client includes a reasonable message in the exception.

timsavage commented:
> Please make sure CI is passing

This is the first time it's run with the full CI suite. I've executed the scripts mentioned (although one was missing numpy), and committed the newly generated files.

It has been a challenge to update as much of this code was written at the end of last year, with the requirements for new integrations (and Python releases) having changed since this was first submitted.

edenhaus reviewed (commented):
Which version of `pydantic` is your lib using?

timsavage commented on homeassistant/components/combined_energy/coordinator.py:
```diff
@@ -0,0 +1,135 @@
+        return await self.api.communication_status()
+
+
+class CombinedEnergyLogSessionService(CombinedEnergyDataService[None]):
```
This is a limitation of the underlying API it requires being poked periodically to keep fetching data. This cannot be removed.

edenhaus replied:
But can we not only call the function, when the coordinator was not getting any data back?

timsavage replied:
No. If the log session is not periodically called, you get gaps in the readings data. This took a lot of trial and error to get working correctly and is the same behaviour observed within the dashboard that combined energy provides.

timsavage commented:
> Which version of `pydantic` is your lib using?

Is not tied to any specific version.

timsavage committed:
Remove CombinedEnergyConnectivityDataService status service, only used by binary sensor

edenhaus commented on homeassistant/components/combined_energy/coordinator.py:
```diff
@@ -0,0 +1,135 @@
+    async def update_data(self) -> ConnectionStatus:
+        """Update data."""
+        return await self.api.communication_status()
+
```
As binary sensor are not part of this PR, please move it to the binary sensor PR.

edenhaus commented on homeassistant/components/combined_energy/coordinator.py:
```diff
@@ -0,0 +1,123 @@
+
+    def __init__(self, hass: HomeAssistant, api: CombinedEnergy) -> None:
+        """Initialize the data service."""
+        super().__init__(hass, api)
```
```suggestion
        super().__init__(hass, api, READINGS_UPDATE_DELAY)
```

edenhaus commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,529 @@
+    installation: Installation = hass.data[DOMAIN][entry.entry_id][DATA_INSTALLATION]
+
+    # Initialise services
+    log_session = CombinedEnergyLogSessionService(hass, api)
```
Please move the creation of the coordinator to `__init__.py` and call `async_config_entry_first_refresh`

edenhaus commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,529 @@
+    ) -> None:
+        """Initialise Readings Sensor."""
+        super().__init__(data_service.coordinator)
+
```
```suggestion
        super().__init__(coordinator)
```

edenhaus commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,529 @@
+class CombinedEnergyReadingsSensor(CoordinatorEntity, SensorEntity):
+    """Representation of a Combined Energy API reading energy sensor."""
+
+    data_service: CombinedEnergyReadingsDataService
```
```suggestion
```

edenhaus commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,529 @@
+        super().__init__(data_service.coordinator)
+
+        self.device_id = device.device_id
+        self.data_service = data_service
```
```suggestion
```

edenhaus commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,529 @@
+    data_service: CombinedEnergyReadingsDataService
+    entity_description: SensorEntityDescription
+
+    native_value_rounding: int = 2
```
Please use [`suggested_display_precision`](https://developers.home-assistant.io/docs/core/entity/sensor) instead so the user can change it if needed.

edenhaus commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,529 @@
+class GenericSensor(CombinedEnergyReadingsSensor):
+    """Sensor that returns the last value of a sequence of readings."""
+
+    def _to_native_value(self, raw_value: Any) -> float:
```
I would suggest using something like https://github.com/home-assistant/core/blob/dev/homeassistant/components/august/sensor.py#L272

edenhaus commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,529 @@
+    def __init__(
+        self,
+        device: Device,
+        device_info: DeviceInfo,
```
```suggestion
```

edenhaus commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,529 @@
+        self.entity_description = description
+
+        self._attr_name = f"{device.display_name} {description.name}"
+        self._attr_device_info = device_info
+        self._attr_unique_id = (
+            f"install_{self.data_service.api.installation_id}-"
+            f"device_{device.device_id}-"
+            f"{description.key}"
+        )
```
```suggestion
        identifier = f"install_{self.api.installation_id}-device_{device.device_id}"
        self._attr_device_info = DeviceInfo(
            identifiers={(DOMAIN, identifier)},
            manufacturer=device.device_manufacturer,
            model=device.device_model_name,
            name=device.display_name,
        )
        self._attr_unique_id = identifier + description.key
```

timsavage commented:
These changes will take a few days to work though.

timsavage committed:
Refactor to flatten dataservice into a subclass of a coordinator and greatly simplify the code to match more modern patterns

timsavage committed:
Refactor sensor to simplify the generation of sensors to a function rather than a factory class. Updated to suit newer HASS patterns

timsavage commented:
Running up the changes in my local instance to test everything is working

edenhaus commented:
@timsavage Please mark the PR as "Ready for review", when you think it should be reviewed again.
A drafted PR will be considered as still in progress and not ready for a review.

edenhaus commented:
Please don't merge each day the dev into this PR. This is not required

edenhaus commented on homeassistant/components/combined_energy/config_flow.py:
```diff
@@ -0,0 +1,105 @@
+            )
+            if can_connect:
+                return self.async_create_entry(
+                    title=user_input.get(CONF_NAME, DEFAULT_NAME),
```
```suggestion
                    title=user_input[CONF_NAME],
```

edenhaus commented on homeassistant/components/combined_energy/coordinator.py:
```diff
@@ -0,0 +1,101 @@
+_T = TypeVar("_T")
+
+
+class CombinedEnergyCoordinator(DataUpdateCoordinator[_T]):
```
```suggestion
class _CombinedEnergyCoordinator(DataUpdateCoordinator[_T]):
```

To make it protected

edenhaus replied:
Can I have feedback on why you did not implement the suggestion?

timsavage replied:
Had a bad rebase, and had to unpick the mess a couple of changes where missed.

edenhaus commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,485 @@
+        SENSOR_DESCRIPTIONS_GENERIC_CONSUMER
+        + [
+            SensorEntityDescription(
+                key="available_energy",
```
All the keys of the water heater are not correct. Please review them and correct them.

timsavage replied:
These are correct.

timsavage replied:
The keys are used to map to entries in the combined energy data.

edenhaus replied:
I would suggest something like `available_liters` as the unit is liters. I can't imagine how energy can be described in liters

timsavage replied:
I can't change this, this is how it is represented in the API. The Unit is correct.

edenhaus commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,485 @@
+                device_class=SensorDeviceClass.TEMPERATURE,
+            ),
+            SensorEntityDescription(
+                key="temp_sensor1",
```
Please explain the value of the sensors 1 to 6? Why can't we name them better?

timsavage replied:
This is the named used in the combined energy API. I can't change it.

edenhaus replied:
What temperature are these sensors describing?

For the user it would be better in my opinion if the sensor has a descriptive name

timsavage replied:
There are up to 6 sensors with 1 at the bottom to n at the top. 1-6 actually makes sense.

edenhaus commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,485 @@
+        return int(round(raw_value[-1], 0))
+
+
+# Map of common device classes to specific sensor types
+SENSOR_TYPE_MAP: dict[
+    SensorDeviceClass | str | None, type[CombinedEnergyReadingsSensor]
+] = {
+    SensorDeviceClass.ENERGY: EnergySensor,
+    SensorDeviceClass.POWER: PowerSensor,
+    SensorDeviceClass.WATER: WaterVolumeSensor,
+    SensorDeviceClass.POWER_FACTOR: PowerFactorSensor,
+}
```
```suggestion
```

Instead of the map please use something similar to https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifi/sensor.py#L118

edenhaus commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,485 @@
+        self.entity_description = description
+
+        identifier = (
+            f"install_{self.coordinator.api.installation_id}-device_{device.device_id}"
```
Why are you extending the identifier by `install` and `device`?

timsavage replied:
I have no guarantee that device_id is globally unique, I do however have a guarantee that a combination of installation_id and device id is.

edenhaus replied:
```suggestion
            f"{self.coordinator.api.installation_id}_{device.device_id}"
```

edenhaus replied:
Why did you only mark it as resolved without implementing the suggestion or leaving any comments?

timsavage commented:
Seems rebase has barfed all over this PR.

timsavage committed:
Recovered from re-base mishap

edenhaus commented on homeassistant/components/combined_energy/config_flow.py:
```diff
@@ -0,0 +1,105 @@
+                )
+
+        else:
+            user_input = {
+                CONF_NAME: DEFAULT_NAME,
+                CONF_USERNAME: "",
+                CONF_PASSWORD: "",
+                CONF_INSTALLATION_ID: "",
+            }
```
```suggestion
            user_input = {
                CONF_NAME: DEFAULT_NAME,
            }
```

edenhaus commented on homeassistant/components/combined_energy/manifest.json:
```diff
@@ -0,0 +1,11 @@
+  "documentation": "https://www.home-assistant.io/integrations/combined_energy",
+  "integration_type": "hub",
+  "iot_class": "cloud_polling",
+  "quality_scale": "silver",
```
Please remove the quality scale for now. You can add it in a follow up PR later on to keep this PR as small as possible
```suggestion
```

edenhaus commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,524 @@
+    return int(round(raw_value[-1], precision))
+
+
+@dataclass
+class CombinedEnergySensorEntityDescriptionMixin:
+    """A class that describes sensor entities."""
+
+    native_value_fn: Callable[[Any, int | None], float | int]
+
+
+@dataclass
+class CombinedEnergySensorEntityDescription(
+    SensorEntityDescription,
+    CombinedEnergySensorEntityDescriptionMixin,
+):
+    """Class describing combined energy sensor entity."""
```
```suggestion
@dataclass(kw_only = True)
class CombinedEnergySensorEntityDescription(SensorEntityDescription):
    """Class describing combined energy sensor entity."""

    native_value_fn: Callable[[Any, int | None], float | int]
```

edenhaus commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,524 @@
+
+    entity_description: CombinedEnergySensorEntityDescription
+    _attr_has_entity_name = True
+    _attr_name = None
```
`_attr_name = None` should only used by the primary entity per device, but currently you are using it for all sensors

timsavage committed:
Update homeassistant/components/combined_energy/config_flow.py

Co-authored-by: Robert Resch <robert@resch.dev>

timsavage committed:
Update homeassistant/components/combined_energy/config_flow.py

Co-authored-by: Robert Resch <robert@resch.dev>

timsavage committed:
Update homeassistant/components/combined_energy/config_flow.py

Co-authored-by: Robert Resch <robert@resch.dev>

timsavage committed:
Update homeassistant/components/combined_energy/manifest.json

Co-authored-by: Robert Resch <robert@resch.dev>

timsavage committed:
Update homeassistant/components/combined_energy/sensor.py

Co-authored-by: Robert Resch <robert@resch.dev>

timsavage committed:
Remove _attr_name = None

timsavage committed:
Update flow tests to work with suggested changes to flow.

timsavage committed:
Update homeassistant/components/combined_energy/sensor.py

Co-authored-by: Robert Resch <robert@resch.dev>

timsavage committed:
Make _CombinedEnergyCoordinator "protected"

edenhaus reviewed (changes requested):
Please mark sure CI is passing.
Some lines need to be formated correctly

timsavage committed:
Re-run ruff over files

timsavage committed:
Merge branch 'dev' into feature/add-combined-energy-hub-integration

edenhaus commented on homeassistant/components/combined_energy/strings.json:
```diff
@@ -0,0 +1,127 @@
+  "entity": {
+    "sensor": {
+      "energy_consumed": {
+        "name": "Energy Consumption"
```
```suggestion
        "name": "Energy consumption"
```

Only the first letter of the first word should be upper case. Please adopt the others

edenhaus commented on homeassistant/components/combined_energy/sensor.py:
```diff
@@ -0,0 +1,513 @@
+    @property
+    def available(self) -> bool:
+        """Indicate if the entity is available."""
+        return self._raw_value is not None
```
```suggestion
        return super().available and self._raw_value is not None
```

timsavage committed:
Assume user input contains CONF_NAME

timsavage commented on homeassistant/components/combined_energy/coordinator.py:
```diff
@@ -0,0 +1,101 @@
+        return self.data
+
+
+class CombinedEnergyLogSessionCoordinator(_CombinedEnergyCoordinator[None]):
+    """Triggers a log session refresh event keep readings data flowing.
+
+    If this is not done periodically, the log session will expire and
+    readings data stops being returned.
+    """
+
+    def __init__(self, hass: HomeAssistant, api: CombinedEnergy) -> None:
+        """Initialize coordinator."""
+        super().__init__(hass, api, LOG_SESSION_REFRESH_DELAY)
+        self.async_add_listener(self.update_listener)
+
+    @staticmethod
+    def update_listener() -> None:
+        """Log that the session has been restarted."""
+        LOGGER.debug("Log session has been restarted")
+
+    async def _update_data(self) -> None:
+        """Update data."""
+        await self.api.start_log_session()
```
We have already covered this in a previous review.  It is required, the linked code calls that end point on the _first_ call, but does not refresh the log session.

edenhaus replied:
The library can easily be improved to check on each call, how long it has passed since the last log_session call was made, and make the call if it is required.

From the docs the function should do that exactly:
```python
 async def _check_for_log_session_restart(self) -> None:
        """Check if a start log session message is required.
```

After finding this function in the library and checking the library, the functionality of the start_log_session should be there and you should not use a coordinator in a hacky way (subscribe itself as listener) for it.

> We have already covered this in a previous review.
The previous review is obsolete as I find the function and the coordinator performs the call even when no entities are enabled (The entities normally add a listener to the coordinator) as you add the coordinator itself as a listener. The last one is not allowed

Please move the functionality of this coordinator to the library.

timsavage committed:
Adjust casing of sensor names

timsavage committed:
Include super().available as well in availability check

timsavage committed:
Remove un-necessary assignment to self.data.

This is handled by the super class anyway.

timsavage committed:
Fixed tests

A suggested change causes tests to fail.

timsavage commented:
Didn't get the stale notification this change is still in progress
